modules/battery_manager/ajax/add_or_edit_entry.php

```diff
@@ -0,0 +1,207 @@
+<?php
+/**
+ * Battery Manager entry adder and editer
+ *
+ * Checks Test Battery for duplicates
+ * Handles insertions and updates in the Test Battery
+ *
+ * PHP Version 7
+ *
+ * @category Loris
+ * @package  Battery_Manager
+ * @author   Victoria Foing <victoria.foing@mcin.ca>
+ * @license  Loris license
+ * @link     https://github.com/aces/Loris-Trunk
+ */
+
+
+// Determine which action is called
+if (isset($_GET['action'])) {
+    $action = $_GET['action'];
+    if ($action == "checkForDuplicate") {
+        echo checkForDuplicate();
+    } else if ($action == "add") {
+        addEntry();
+    } else if ($action == "edit") {
+        editEntry();
+    }
+}
+
+/**
+ * Return duplicate of entry if it exists and null otherwise
+ *
+ * @return json object or null
+ */
+function checkForDuplicate()
+{
+    $db =& \Database::singleton();
+
+    // Retrieve values entered by user
+    $form_data = getFormData();
+
+    // Build SQL query based on values entered by user
+    $query     = " SELECT
+                   ID,
+                   Test_name,
+                   AgeMinDays,
+                   AgeMaxDays,
+                   Active,
+                   Stage,
+                   SubprojectID,
+                   Visit_label,
+                   CenterID,
+                   firstVisit,
+                   instr_order FROM test_battery ";
+    $i         = 0;
+    $connector = "WHERE ";
+    // Iterate through values entered by user
+    foreach ($form_data as $key => $value) {
+        if ($i > 0) {
+            $connector = "AND ";
+        }
+        if (isset($value) || $value !== null) {
+            $query .= $connector . $key . " = :" . $key . " ";
```

Comment:
This is dangerous and should be using a prepared statement.

Reply:
to find a duplicate of an entry, we have to build an SQL query that looks at all the values of the entry. the for loop goes through all the values in the entry and determines if they have been filled out or not. if a value is filled out it adds " $key = :$key ". if a value is not filled out, it is NULL so the SQL query must contain " $key IS NULL " instead of " $key = :$key " and the $key must be removed from the parameter array.
for example, if I add an entry with the following values:
Test_name: bmi, AgeMinDays: 0, AgeMaxDays: 99999, Stage: Visit, and I do not fill out SubprojectID, Visit_label, CenterID, firstVisit, instr_order. 
The SQL query is "SELECT * FROM test_battery WHERE Test_name = :Test_name AND AgeMinDays =:AgeMinDays AND AgeMaxDays = :AgeMaxDays AND Stage = :Stage AND SubprojectID IS NULL AND Visit_label IS NULL AND CenterID is NULL AND firstVisit IS NULL AND instr_order IS NULL" with parameters ['Test_name' => 'bmi', 'AgeMinDays' => '0', 'AgeMaxDays' => '99999', 'Stage' => 'Visit']